File: News.java
import java.util.*;
import java.io.*;

public class News {

 public static final int DEFAULT_SIZE = 250;

 public static void main( String[] args ) {
  try {
   Scanner textSC = new Scanner(new File("newsCorpora_full.txt"));
   Scanner csvSC = new Scanner(new File("newsCorpora.csv"));
   int[] invalidIndices = {6, 82, 84, 86, 94};
   ArrayList<String> b = new ArrayList<String>();
   ArrayList<String> t = new ArrayList<String>();
   ArrayList<String> e = new ArrayList<String>();
   ArrayList<String> m = new ArrayList<String>();

   // extract news
   StringBuilder builder = new StringBuilder();
   int id = -1;
   int previousId = 0;
   while (textSC.hasNextLine()) {
    String line = textSC.nextLine();
    // beginning of a news
    if (line.indexOf(">>>>") >= 0) {
     builder = new StringBuilder();
     id = Character.getNumericValue(line.charAt(line.length()-1));
    }
    else if (line.indexOf("<<<<") >= 0) {
     String news = builder.toString();
     if (news.length() == 0) continue;
     String[] newsInfo = findInfo(csvSC, id, previousId);
     previousId = id;
     id = -1;
     if (newsInfo.length < 8) continue;
     String type = newsInfo[4];
     if ( type.equals("b") && b.size() < DEFAULT_SIZE ) {
      b.add(news);
     }
     else if ( type.equals("t") && t.size() < DEFAULT_SIZE ) {
      t.add(news);
     }
     else if ( type.equals("e") && e.size() < DEFAULT_SIZE ) {
      e.add(news);
     }
     else if ( type.equals("m") && m.size() < DEFAULT_SIZE ) {
      m.add(news);
     }
     else {
      System.err.println("unrecognized type: " + type);
     }
    }
    else if (!contains(invalidIndices, id)) {
     builder.append(line + "\n");
    }
    if (b.size() >= DEFAULT_SIZE && t.size() >= DEFAULT_SIZE && e.size() >= DEFAULT_SIZE && m.size() >= DEFAULT_SIZE)
     break;
   }
   FileWriter writer = new FileWriter(new File("shortlist.txt"));
   printArr(writer, b, "b");
   printArr(writer, t, "t");
   printArr(writer, e, "e");
   printArr(writer, m, "m");
   writer.flush();
   writer.close();
   textSC.close();
  } catch (Exception e) {
   e.printStackTrace();
  }
 }


 public static String[] findInfo(Scanner sc, int newsId, int previousId) {

  String line = "";
  for (int i = previousId ; i < newsId ; i++) {
   line = sc.nextLine();
   if ( !sc.hasNextLine() ) break;
  }
  return line.split("\t");
 }


 public static boolean contains( int[] arr, int x ) {
  for ( int y : arr ) {
   if ( y == x ) return true;
  }
  return false;
 }


 public static void printArr(FileWriter writer, ArrayList<String> arr, String title) {
  try {
   writer.write("***** " + title + " *****\n");
   for ( int i = 0 ; i < arr.size() ; i++ ) {
    writer.write(">>>>" + (i+1) + "\n" + arr.get(i) + "<<<<" + (i+1) + "\n");
   }
   writer.write("*****\n");
  } catch (Exception e) {
   e.printStackTrace();
  }
 }
}

File: news.py
def write_result(size):
	text_scanner = open("newsCorpora_full.txt", "r")
	csv_scanner = open("newsCorpora.csv", "r")
	invalid_indices = [6, 82, 84, 86, 94]
	b = []
	t = []
	e = []
	m = []

	# the id of the current news
	news_id = -1

	lines = text_scanner.readlines()
	info_lines = csv_scanner.readlines()
	current_news = ""
	for line in lines:
		# beginning of a news
		if ">>>>" in line:
			# remove leading and trailing whitespace
			stripped = line.strip()
			news_id = int(line[4:])
			current_news = ""
		# end of a news
		elif "<<<<" in line:
			if len(current_news) == 0:
				continue
			# extract information from csv file based on the news_id
			# the type is on the 5th column
			news_type = info_lines[news_id].split("\t")[4]
			# update the ids
			news_id = -1
			# add the news to the appropriate list
			add_news(current_news, news_type, size, b, t, e, m)
		elif news_id not in invalid_indices:
			current_news = current_news + line

		if len(b) >= size and len(t) >= size and len(e) >= size and len(m) >= size:
			break

	file_writer = open("sl" + str(size * 4) + ".txt", "w+")
	print_list(file_writer, b, "b")
	print_list(file_writer, t, "t")
	print_list(file_writer, e, "e")
	print_list(file_writer, m, "m")

	text_scanner.close()
	csv_scanner.close()
	file_writer.close()


def add_news(news, news_type, size, b, t, e, m):
	if news_type == "b":
		if len(b) < size:
			b.append(news)
	elif news_type == "t":
		if len(t) < size:
			t.append(news)
	elif news_type == "e":
		if len(e) < size:
			e.append(news)
	elif news_type == "m":
		if len(m) < size:
			m.append(news)
	else:
		print("unrecognized type " + news_type)


def print_list(writer, list, title):
	writer.write("***** " + title + " *****\n")
	for i in range(len(list)):
		writer.write(">>>>" + str(i+1) + "\n" + list[i] + "<<<<" + str(i+1) + "\n");
	writer.write("*****\n");


if __name__ == '__main__':
    write_result(180)

